[CBRD-26239] refine synchronization in cubpl::session

hyunikn opened the pull request:
<http://jira.cubrid.org/browse/CBRD-26239>

### Purpose

pl session 의 상태 - stack, connection, cursor - 를 동기화 하기 위해 mutex 와 condition variable 을 사용하고 있다. 
동기화 관련 코드를 좀더 효율적이고 가독성 높게 수정한다. 

### Implementation

- m_mutex 를 세 개의 mutex 로 분리 : 서로 독립적인 세 가지 종류의 상태를 각각의 mutex 로 동기화 하는 것이 바람직
  - m_mutex_stack
  - m_mutex_connection
  - m_mutex_cursor
- m_cond_var 를 두 개의 condition variable 로 분리: 서로 다른 두 개의 조건을 각각의 condition variable 로 분리하는 것이 바람직.
  - m_cond_target_stack_at_top
  - m_cond_pl_session_done
- session_notify_pl_task_completion() 함수 삭제: session::pop_and_destroy_stack() 에서 하고 있는 notify_all 과 중복
- unique_lock 보다 lightweight 한 lock_guard 를 가급적 사용
  - unlock() 할 필요 있거나 condition variable 과 함께 사용할 때는 unique_lock 을 사용해야 한다고 함.
- stack 관련 로직 refine
- 코드 정리
  - m_is_running, m_is_interrupted 필드 삭제: 동일 정보를 얻을 수 있는 다른 필드가 있음
    - m_is_running 은 m_stack_idx >= 0 와 동일
    - m_is_interrupted 는 m_interrupt_id != 0 과 동일

### Remarks

N/A

hyunikn committed:
minor: remove m_is_running from cubpl::session

hyunikn committed:
separate m_mutex into three mutexes - m_mutex_stack, m_mutex_connection, m_mutex_cursor

hyunikn committed:
minor: remove m_is_interrupted from cubpl::session

hyunikn committed:
. separate m_cond_var into two condition variables - m_cond_target_stack_at_top and m_cond_pl_session_done
. remove a function session_notify_pl_task_completion()
. refine stack manipulation logic

hyunikn committed:
Merge branch 'develop' into cbrd-26239-pl-session-cond-var

hyunikn committed:
code style

hyunikn commented:
/run all

hyunikn committed:
remove cubpl::session::notify_waiting_stacks because it is redundant

hyunikn committed:
minor fix
. merge session::destroy_cursor_internal to its call site
. simplify boolean returning code

hyunikn commented:
/run all

hyunikn commented on src/session/session.c:
```diff
@@ -3362,7 +3349,7 @@ session_stop_attached_threads (THREAD_ENTRY * thread_p, void *session_arg, bool
 	{
 	  session->pl_session_p->set_interrupt (er_errid ());
-	  session->pl_session_p->wait_for_interrupt ();
+	  session->pl_session_p->wait_until_pl_session_done ();
```
함수 이름만 바꾸었습니다.

hyunikn commented on src/sp/pl_session.cpp:
```diff
@@ -110,32 +109,35 @@ namespace cubpl
     execution_stack *stack = new execution_stack (thread_p);
     if (stack)
       {
-	m_stack_map [stack->get_id ()] = stack;
```
stack_id 변수를 사용할 수 있는 위치로 옮겼습니다.

hyunikn commented on src/sp/pl_session.cpp:
```diff
@@ -158,60 +158,80 @@ namespace cubpl
   void
   session::pop_and_destroy_stack (const PL_STACK_ID sid)
   {
-    if (m_stack_idx == -1)
```
이런 일이 발생할 일이 없다고 생각해서 지웠습니다. 
PR commit 172 라인에서 assert (m_stack_idx >= 0); 합니다.

ctshim replied:
172 라인에서 `assert (m_stack_idx >= 0);`를 보장하지만 바로 그 아래
`m_stack_idx--;`에 의해 `-1`이 되는 것도 당연히 발생할 수 있는 개념입니다.
착각을 하신게 아닌지요?

hyunikn replied:
stack push 된 만큼만 pop 될 것이기 때문입니다. 
m_stack_idx--; 를 거쳐 -1 이 되면 스택이 다시 쌓이기 (m_stack_idx 가 다시 >= 0) 
전까지는 더이상 pop 될 stack 이 없습니다.

hyunikn commented on src/sp/pl_session.cpp:
```diff
@@ -158,60 +158,80 @@ namespace cubpl
 
-    if (pred ())
-      {
-	if (m_stack_idx > -1)
```
검사하지 않고 assert 로 바꾸었습니다.

hyunikn commented on src/sp/pl_session.cpp:
```diff
@@ -539,28 +509,21 @@ namespace cubpl
   session::is_session_cursor (QUERY_ID query_id)
   {
-    std::lock_guard<std::mutex> ulock (m_mutex);
-    if (m_session_cursors.find (query_id) != m_session_cursors.end ())
```
동작에 변화 없는 코드 정리 입니다.

hyunikn commented on src/sp/pl_session.hpp:
```diff
@@ -150,7 +149,7 @@ namespace cubpl
       std::string get_interrupt_msg ();
       void clear_interrupt ();
 
-      void wait_for_interrupt ();
```
동작에 변화 없는 이름 변경입니다.

hyunikn commented on src/sp/pl_session.hpp:
```diff
@@ -206,12 +208,9 @@ namespace cubpl
       int m_stack_idx;
 
       // interrupt
-      bool m_is_interrupted;
```
m_interrupt_id != 0 과 같은 값이므로 삭제했습니다.

hyunikn commented on src/sp/pl_session.hpp:
```diff
@@ -206,12 +208,9 @@ namespace cubpl
       int m_interrupt_id;
       std::string m_interrupt_msg;
 
-      bool m_is_running;
```
m_stack_idx >= 0 와 같은 값이므로 삭제 했습니다.

hyunikn commented on src/sp/pl_session.hpp:
```diff
@@ -178,8 +177,11 @@ namespace cubpl
       void destroy_all_cursors ();
       void destroy_pl_context_jvm ();
 
-      std::mutex m_mutex;
```
mutex 는 3개로 condition variable 은 2개로 분리했습니다.

hyunikn commented:
/run all

ctshim commented:
하나의 뮤텍스를 사용하던 것을 3개로 나누어서 세분화 하는 것은 좋은 생각 같습니다.
다만, 한 가지 염려가 되는 것은  3개로 나눈 만큼
하나의 mutex를 잡고 있는 상태에서 다른 mutex를 잡으려고 하는 경우가 혹시 있는지 확인이 필요해 보입니다.
dead-lock 이슈나, 사실상 하나만 있어도 되는  것을 3개로 늘린 결과는 아닌지에 대해서도 고민이 필요한데, 이 부분 검토를 하셨던 건지 리뷰 전에 미리 여쭤 봅니다.^^

ctshim commented on src/connection/server_support.c:
```diff
@@ -2778,8 +2778,6 @@ css_server_task::execute (context_type &thread_ref)
   pthread_mutex_lock (&thread_ref.tran_index_lock);
   (void) css_internal_request_handler (thread_ref, m_conn);
 
-  (void) session_notify_pl_task_completion (session_p);
-
```
`session_notify_pl_task_completion() 함수 삭제: session::pop_and_destroy_stack() 에서 하고 있는 notify_all 과 중복`

잘 이해를 못했습니다. 
호출 시점의 차이가 있는데,  이 함수(execute)가 종료되는 즉시 pop_and_destroy_stack()가 호출되는 건가요?

hyunikn replied:
pop_and_destroy_stack ()은 이 실행될 때 call stack 의 저 아래 쪽에 execute() 이 있습니다. 
pop_and_destroy_stack () 안에서 필요한 notify_all 이 모두 불리워 진다고 보고 execute 에서는 지웠습니다.

hyunikn replied:
push 된 stack이 모두 pop 되는지는 확인해 보겠습니다. 그렇지 않다면 문제가 될 수 있겠습니다.

hyunikn replied:
pop_and_destroy_stack () 이 불리지 않고 execute() 이 리턴 한다는 것은 그 문제 뿐만 아니라 
memory leak 을 의미하는 것이기도 합니다. 
pop_and_destroy_stack () 은 execution_stack 의 소멸자에서 호출되기 때문입니다.

hyunikn replied:
execution_stack 은 cubpl::executor 의 생성자에서 만들어지고 소멸자에서 소멸됩니다. 현재 구현에서는 만들어진 execution_stack 이 소멸되지 않는 경우는 없는 것으로 판단됩니다. 왜냐하면 cubpl::executor 이 new로 만들어지는 일이 없고 언제나 로컬 변수로만 사용되기 때문입니다.

hyunikn commented:
> 하나의 뮤텍스를 사용하던 것을 3개로 나누어서 세분화 하는 것은 좋은 생각 같습니다. 다만, 한 가지 염려가 되는 것은 3개로 나눈 만큼 하나의 mutex를 잡고 있는 상태에서 다른 mutex를 잡으려고 하는 경우가 혹시 있는지 확인이 필요해 보입니다. dead-lock 이슈나, 사실상 하나만 있어도 되는 것을 3개로 늘린 결과는 아닌지에 대해서도 고민이 필요한데, 이 부분 검토를 하셨던 건지 리뷰 전에 미리 여쭤 봅니다.^^

- 1개이던 것을 3개로 나눈다고 해서 없던 dead-lock 이 생기지는 않을 것입니다. 오히려 있던 dead-lock이 없어질 수는 있을 겁니다. 
- 테스트 결과 is_running() 에 없던 lock 잡기를 추가했더니 한 thread 에서 동일 mutex lock 을 잡는 상황이 생겨 dead lock 이 발생했습니다. 수정하겠습니다.
- 사실 1개로 두어도 동작에는 문제 없으나 독립적으로 접근되는 다른 종류의 자원을 보호하기 위해 동일 mutex 를 쓰게 되면 더 오래 기다리는 경우가 생겨 성능에 약간이라도 손해가 있을 것입니다. 
- 참고로, 정상 동작의 경우 한 시점에 active 한 thread 는 오직 하나이므로 사실 mutex 나 condition variable 을 사용한 동기화가 필요하지 않습니다. 그러나, 에러의 경우 동기화가 필요합니다. 예를 들어, 재귀 호출에 의해 여러 thread 가 다음 번 호출 결과를 기다리며 block 되어 있다가 CAS 가 죽어 한꺼번에 이 thread 들이 실행을 재개하면 동기화가 필요해 집니다. 이 때 여러 thread 가 서로 다른 종류의 자원을 접근할 수 있겠습니다.

hyunikn committed:
. remove an unused method cubpl::session::get_depth()
. take the cubpl:session as an argument of the constructor of cubpl::execution_stack

hyunikn commented:
/run all

hyunikn committed:
minor rename - cubpl::session::is_running --> cubpl::session::is_sp_running

hyunikn committed:
Merge branch 'develop' into cbrd-26239-pl-session-cond-var

hyunikn commented:
/run all

hyunikn commented on src/communication/network_interface_sr.c:
```diff
@@ -167,7 +167,7 @@ stran_server_commit_internal (THREAD_ENTRY * thread_p, unsigned int rid, bool re
     {
       PL_SESSION *session = cubpl::get_session ();
-      if (!session || session->is_running () == false)
+      if (!session || session->is_sp_running () == false)
```
동작 변경 없는 메소드 이름 변경 입니다. 
is_running 이 다른 클래스에서도 많이 쓰여 구별을 위해 바꾸었습니다.

hyunikn committed:
use lock_guard in as many cases as possible which is more lightweight than unique_lock

hyunikn commented:
/run all

hyunikn commented on src/sp/pl_execution_stack_context.cpp:
```diff
@@ -39,13 +39,8 @@ namespace cubpl
     m_tid = logtb_find_current_tranid (thread_p);
     m_is_running = false;
-
-    session *sess = get_session ();
```
get_session() 내부에서 cubpl::session::is_sp_running() 을 호출에 도달하고 lock 문제가 발생하여 
get_session() 을 호출하지 않고 인자로 받아오도록 바꾸었습니다.

hyunikn committed:
minor fix: restore clear_interrupt() for the first execution stack

hyunikn committed:
minor: restore the original position of session::is_thread_involved() in the file.

hyunikn commented on src/sp/pl_session.hpp:
```diff
@@ -149,12 +148,10 @@ namespace cubpl
+      void wait_until_pl_session_done ();
       void set_local_error_for_interrupt (); // set interrupt on thread local error manager
 
-      int get_depth ();
```
get_depth() 는 사용되는 곳이 없어 지웠습니다.

hyunikn commented:
/run all

hyunikn commented:
shell test 의 결과 아래와 같이 11 개의 실패가 발생했습니다. 
```
shell/_04_misc/_04_index/itrack_10001/cases/itrack_10001.sh
shell/_04_misc/_04_index/itrack_10005/cases/itrack_10005.sh
shell/_06_issues/_11_1h/bug_bts_5492/cases/bug_bts_5492.sh
shell/_06_issues/_12_2h/bug_bts_9521_1/cases/bug_bts_9521_1.sh
shell/_06_issues/_12_2h/bug_bts_9601/cases/bug_bts_9601.sh
shell/_06_issues/_24_2h/cbrd_25546/cases/cbrd_25546.sh
shell/_06_issues/_25_1h/cbrd_26020/cases/cbrd_26020.sh
shell/_08_shard/_17_shard_cas_log/cases/_17_shard_cas_log.sh
shell/_10_plcsql/cbrd_25619/cases/cbrd_25619.sh
shell/_35_cherry/issue_21506_online_index/cbrd_22477_unload/cases/cbrd_22477_unload.sh
shell/_35_cherry/issue_21654_server_side_loaddb/loaddb_CS/_35_cherry/issue_21506_online_index/cbrd_22477_unload/cases/cbrd_22477_unload.sh

```
이 중에서 기반이 된 develop branch commit b8505c4 에서 발생한 실패
(https://app.circleci.com/pipelines/github/CUBRID/cubrid/21287/workflows/758f5a8d-c98d-4dc1-a504-3b642a74ec0d/jobs/100943/tests) 에 비해서 추가로 발생한 것은 아래 다섯 건입니다. 
```
shell/_06_issues/_12_2h/bug_bts_9521_1/cases/bug_bts_9521_1.sh  - 이미 unstable 로 리포트 됨
shell/_06_issues/_12_2h/bug_bts_9601/cases/bug_bts_9601.sh  - 개인 개발 VM에서는 성공
shell/_06_issues/_24_2h/cbrd_25546/cases/cbrd_25546.sh  - 이미 unstable 로 리포트 됨
shell/_10_plcsql/cbrd_25619/cases/cbrd_25619.sh  - 이미 unstable 로 리포트 됨
shell/_35_cherry/issue_21506_online_index/cbrd_22477_unload/cases/cbrd_22477_unload.sh  - 이미 unstable 로 리포트 됨

```
다섯 건 모두 unstable 로 추정됩니다.

ctshim commented on src/sp/pl_session.cpp:
```diff
@@ -374,35 +366,21 @@ namespace cubpl
   {
-    m_is_interrupted = false;
-    m_interrupt_id = NO_ERROR;
+    m_interrupt_id = 0;
```
기존 코드가 더 의도가 명확한 것 같습니다.
```suggestion
    m_interrupt_id = NO_ERROR;
```

hyunikn replied:
원복하겠습니다.

ctshim commented on src/sp/pl_session.cpp:
```diff
@@ -74,9 +76,7 @@ namespace cubpl
-    , m_is_interrupted (false)
-    , m_interrupt_id (NO_ERROR)
-    , m_is_running (false)
+    , m_interrupt_id (0)
```
기존 방식이 더 명확한 것 같습니다.
```suggestion
    , m_interrupt_id (NO_ERROR)
```

hyunikn replied:
네. 원복 하겠습니다. 
if (m_interrupt_id)  이렇게 간단하게 쓰려고 바꾼 것입니다만, 
(이렇게 쓰면 NO_ERROR 값이 나중에라도 혹시 0이 아니게 바뀐다면 문제가 됩니다. ) 
if (m_interrupt_id != NO_ERROR) 가 더 명확하네요.

kangmin5505 commented on src/sp/pl_session.cpp:
```diff
@@ -158,47 +158,47 @@ namespace cubpl
+    assert (m_stack_idx >= 0);
     m_exec_stack[m_stack_idx] = -1;
     m_stack_idx--;
+    m_cond_target_stack_at_top.notify_all();
 
     m_stack_map.erase (sid);
```
대기 중인 threads 에게 `notify_all()` 을 하고 `m_stack_map.erase()` 하는 것보다 함수 마지막에 `ulock.unlock()`; 이후 `notify_all()`을 수행하는 것이  대기하던 threads가 모두 다시 잠드는 일이 발생하지 않기 때문에 더 적절하다고 생각하다고 생각합니다.
하지만 현익님으로부터 큰 차이가 없을 거라고 답변을 받아 comment로만 남깁니다.

hyunikn replied:
unlock() 을 명시적으로 불러준다는 것은 생각해 보지 않았습니다만, 
그렇게 해도 모두 잠드느냐, 아니면 하나 빼고 모두 잠드느냐의 차이 정도라 미미할 것 같습니다. 
반면, 현재 위치가 notify_all() 을 하게 되는 원인에  (m_stack_idx 감소) 가깝다는 점이 있습니다.

hyunikn committed:
minor: restore NO_ERROR as the initial value of cubpl::session::m_interrupt_id

hyunikn commented:
/run all

beyondykk9 commented on src/sp/pl_session.cpp:
```diff
@@ -305,7 +299,7 @@ namespace cubpl
   void
   session::set_interrupt (int reason, std::string msg)
```
std::string msg는 사용이 안되고 있는 것 같습니다. 실제 이 function을 콜하는 쪽에서도 reason만 argument로 넘기고 있습니다.

hyunikn replied:
네. 다시 보니 그렇습니다. 현재는 msg 가 언제나 디폴트 값 빈문자열 ( "" ) 입니다.
다만, cubmethod::handle_method_error() 에 인자로 넘기는 부분이 있습니다. 
향후 사용을 염두에 두고 그냥 두어도 좋을 것 같습니다.

beyondykk9 commented on src/sp/pl_session.cpp:
```diff
@@ -99,43 +99,45 @@ namespace cubpl
-	m_cond_var.notify_all ();
-	return nullptr;
+
+	assert (m_stack_idx >= 0);
```
m_stack_idx는 -1 아래 값을 가질 수 없기 때문에 앞선 if (m_stack_idx == -1) 에 의해 m_stack_idx는 당연 0보다 크거나 같지 않을까요?

hyunikn replied:
네. 바로 그 점을 점검하는 assert 입니다. 
혹시라도 -2, -3 값을 갖는다면 여기서 assertion fail 이 날 것입니다.

hyunikn commented:
@airnet73 @hgryoo  시간 관계 상 리뷰를 여기까지 마치고 merge 하겠습니다. 
혹시 나중에라도 문제가 보이시면 알려 주십시오.

hyunikn commented:
/run sql
